feat(dashboard): Open linked document instead of list when count is 1

cogk opened the pull request:
## Description

When a dashboard link has only one document, open the document instead of the list when clicking on the link.

This is done by adding the `href` attribute to the link, and skipping any `frappe.set_route` calls.
Additional attributes `data-doctype` and `data-name` are added to enable the link preview popover, as shown in the picture below.

![image](https://github.com/frappe/frappe/assets/10946971/87791be8-7d6f-4ff2-aa6f-e7b58db345c6)

---

I tried to be forward compatible, in case document-links stop using `set_route` to instead all have `href`s (to allow ctrl/cmd-click to open in a new tab, etc.).

> no-docs

cogk committed:
feat(dashboard): Open linked document instead of list when count is 1

cogk committed:
test(dashboard): Fix test_external_link_count

cogk committed:
test(ui): Add tests for dashboard links

ankush commented:
Personally don't like UI elements doing two different things based on state of data, but this can be nifty for *most* users who will likely have 1 document linked *most* of the time. 


@surajshetty3416 / @shariquerik ?

---

related/alternate way to solve same problem: https://github.com/frappe/frappe/issues/21192

cogk committed:
test(ui): Fix dashboard test

It seems that tests are cross-contaminating themselves by changing the UI state. Maybe clearing localStorage can help?

bosue commented:
> I tried to be forward compatible… to instead all have `hrefs`

♥️

> Personally don't like UI elements doing two different things based on state of data, but this can be nifty for most users who will likely have 1 document linked most of the time.

True. We could mitigate it by not doing resp. showing two completely different things but to bring one and the other in line as much as possible. Even for multiple links, there could be a popover preview of the resulting table. And if there‘s only one of them, we would still show it in a table but expanded.

Wouldn‘t this be a nice UI pattern to have in table / list view just as well, for _any_ kind of doctypes being listed? Hover and you‘ll see a preview of the doc? 😍

bosue commented:
> Even for multiple links, there could be a popover preview of the resulting table. And if there‘s only one of them, we would still show it in a table but expanded.

A table with multiple columns is too bulky. But the idea of a vertical accordion would work: expanded for a single item, condensed for the first, say: five, of multiple items linked.

ankush commented:
Closing for now, refer my first comment for reasons.

bosue commented:
@ankush: Sure we took your comment very serious and (I think) all agree:
Doing completely different things based on state is no good UI pattern.

Question remains whether you would be okay with an alternative approach following my proposal.
But the answer will probably be: it depends if it‘s done right, so we need to figure out what‘s right. 🤷🏻‍♂️

ankush commented:
@bosue I don't have authority on design changes :laughing: 

One of @surajshetty3416 / @netchampfaris / @shariquerik needs to be approve it.
